rewrite-java/src/main/java/org/openrewrite/java/internal/beta/CallMatcher.java

```diff
@@ -0,0 +1,133 @@
+    }
+
+    @AllArgsConstructor(access = AccessLevel.PRIVATE)
+    class AdvancedCallMatcher {
+        CallMatcher matcher;
+
+        public boolean isSelect(Expression expression, Cursor cursor) {
+            assert expression == cursor.getValue() : "expression != cursor.getValue()";
+            J.MethodInvocation maybeMethodInvocation =
+                    cursor.getParentOrThrow().firstEnclosing(J.MethodInvocation.class);
+            return maybeMethodInvocation != null &&
+                    maybeMethodInvocation.getSelect() == expression &&
+                    matcher.matches(maybeMethodInvocation); // Do the matcher.matches(...) last as this can be expensive
+        }
+
+        public boolean isAnyArgument(Expression expression, Cursor cursor) {
+            assert expression == cursor.getValue() : "expression != cursor.getValue()";
+            J stop = cursor.dropParentUntil(v -> v instanceof J.MethodInvocation || v instanceof J.NewClass || v instanceof J.Block).getValue();
+            if (stop instanceof J.Block) {
+                return false;
+            }
+            Expression call = (Expression) stop;
+            return getCallArguments(call).contains(expression) && matcher.matches(call); // Do the matcher.matches(...) last as this can be expensive
+        }
+
+        public boolean isFirstParameter(Expression expression, Cursor cursor) {
+            return isParameter(expression, cursor, 0);
+        }
+
+        public boolean isParameter(Expression expression, Cursor cursor, int parameterIndex) {
+            assert expression == cursor.getValue() : "expression != cursor.getValue()";
+            J stop = cursor.dropParentUntil(v -> v instanceof J.MethodInvocation || v instanceof J.NewClass || v instanceof J.Block).getValue();
+            if (stop instanceof J.Block) {
+                return false;
+            }
+            Expression call = (Expression) stop;
+            List<Expression> arguments = getCallArguments(call);
+            if (parameterIndex >= arguments.size()) {
+                return false;
+            }
+            if (doesMethodHaveVarargs(call)) {
+                int finalParameterIndex = getType(call).getParameterTypes().size() - 1;
+                if (finalParameterIndex == parameterIndex) {
+                    List<Expression> varargs = arguments.subList(finalParameterIndex, arguments.size());
+                    return varargs.contains(expression) &&
+                            matcher.matches(call); // Do the matcher.matches(...) last as this can be expensive
+                }
+            }
+            return arguments.get(parameterIndex) == expression &&
+                    matcher.matches(call); // Do the matcher.matches(...) last as this can be expensive
+        }
+
+        private static boolean doesMethodHaveVarargs(Expression expression) {
+            return getType(expression).hasFlags(Flag.Varargs);
+        }
+
```

Comment:
unfortunately null and unknown types are quite common, I would suggest removing the assertions and adding a `@Nullable` annotation to this method